`RFDETRLarge(resolution=...)` silently falls back to deprecated config because `load_pretrain_weights` lacks positional embedding interpolation

### Search before asking

- [x] I have searched the RF-DETR issues and found no similar bug report.


### Bug


This is a follow-up to #948 and PR #956. PR #956 correctly fixes the architecture-level part of the resolution problem — `positional_encoding_size` is now derived from the user-supplied `resolution` at construction time. 

However, while testing the merged fix on the `develop` branch, I found that the resolution path is still broken end-to-end, and the failure mode on `RFDETRLarge` is particularly subtle because a `try/except` in `RFDETRLarge.__init__` silently swallows the real error and falls back to the deprecated Large config. Users who pass a custom resolution to `RFDETRLarge` could end up training a completely different model.

Below are two connected problems with minimal reproductions.

### Environment

- RF-DETR: develop @ `c80bda9` (post-#956 merge)
- OS: Linux
- Python: 3.11
- PyTorch: 2.6.0+cu124
- GPU: NVIDIA GeForce RTX 3060

### Minimal Reproducible Example

## Problem A — `load_pretrain_weights` lacks positional embedding interpolation

This is the cleanest reproduction. Default-construct the model (so the architecture matches the pretrained checkpoint), then pass `resolution=640` to `train()`:

```python
model = RFDETRLarge()
model.train(
    dataset_dir=str(DATASET_DIR),
    epochs=1,
    batch_size=8,
    grad_accum_steps=2,
    resolution=640,
)
```

```
RuntimeError: Error(s) in loading state_dict for LWDETR:
    size mismatch for backbone.0.encoder.encoder.embeddings.position_embeddings:
    copying a param with shape torch.Size([1, 1937, 384]) from checkpoint,
    the shape in current model is torch.Size([1, 1601, 384]).
```

What happens:

1. `RFDETRLarge()` constructs cleanly with the Large config (patch=16, PE=44, 704×704).
2. `train(resolution=640)` applies the new `resolution` to `model_config`, which (correctly, post-#956) rebuilds the model with PE=40 (1601 tokens).
3. `load_pretrain_weights` then tries to load the pretrained checkpoint (1937 tokens, 44×44) into the new model (1601 tokens, 40×40) and crashes on `load_state_dict`.

The same crash reproduces on `RFDETRSmall(resolution=640)` (1025 vs 1601) and `RFDETRMedium(resolution=640)` (1297 vs 1601), so it's not Large-specific — it's a general gap in `load_pretrain_weights`, which never interpolates positional embeddings between source and target grid sizes.

This is a standard ViT/DINOv2 operation: bicubic-interpolate the checkpoint's PE grid from the source shape (e.g. 44×44) to the target shape (e.g. 40×40) before calling `load_state_dict`. @isaacrob already confirmed in #956 that PE interpolation is acceptable as long as it's not baked into the export graph (interpolate, then trace).

Without this, the user-facing `resolution` parameter is unusable on any pretrained model whose checkpoint PE grid doesn't already match the target.

---

## Problem B — `RFDETRLarge` silently falls back to the deprecated config

This is a more subtle and arguably more dangerous failure. Construct `RFDETRLarge` with a custom resolution and then call `train()` with no resolution argument:

```python
model = RFDETRLarge(resolution=640)
model.train(
    dataset_dir=str(DATASET_DIR),
    epochs=1,
    batch_size=8,
    grad_accum_steps=2,
)
```

```
AssertionError: Backbone requires input shape to be divisible by 56,
but got torch.Size([8, 3, 640, 640])
```

The magic number `56` is the smoking gun. `56 = patch_size × num_windows`. For the **modern** Large config that's `16 × 2 = 32`. For the **deprecated** Large config it's `14 × 4 = 56`. So this error proves the model is being built with the deprecated Large architecture, even though the user wrote `RFDETRLarge(resolution=640)`.

The cause is in `src/rfdetr/variants.py`. `RFDETRLarge.__init__` wraps the modern-config construction in a try/except that falls back to the deprecated config if a specific set of error conditions is met:

```python
class RFDETRLarge(RFDETR):
    @staticmethod
    def _should_fallback_to_deprecated_config(exc: Exception) -> bool:
        message = str(exc).lower()
        if "out of memory" in message:
            return False
        if isinstance(exc, ValueError):
            return "patch_size" in message
        if isinstance(exc, RuntimeError):
            incompatible_state_dict_markers = (
                "error(s) in loading state_dict",
                "size mismatch",
                "missing key(s) in state_dict",
                "unexpected key(s) in state_dict",
            )
            return any(marker in message for marker in incompatible_state_dict_markers)
        return False

    def __init__(self, **kwargs):
        self.init_error = None
        self.is_deprecated = False
        try:
            super().__init__(**kwargs)
        except (ValueError, RuntimeError) as exc:
            if not self._should_fallback_to_deprecated_config(exc):
                raise
            self.init_error = exc
            self.is_deprecated = True
            try:
                super().__init__(**kwargs)
                logger.warning("... Automatically switched to deprecated model configuration ...")
            except Exception:
                raise self.init_error
```

The problem is that **the heuristic assumes the only cause of a state_dict size mismatch is deprecated weights**. That assumption held before #956. After #956, there is a second way to get a size mismatch: the user passes a custom `resolution`, the modern config correctly rebuilds the model with a new positional encoding grid (e.g. PE=40 for `resolution=640`), and then `load_pretrain_weights` crashes loading the default *modern* pretrained checkpoint (PE=44 for 704×704) because positional embeddings aren't interpolated (Problem A above).

Trace for `RFDETRLarge(resolution=640)` on current develop:

1. `super().__init__` tries the modern config (patch=16, num_windows=2).
2. #956 correctly sets `positional_encoding_size = 640 // 16 = 40`. Model is built with 1601 PE tokens.
3. `load_pretrain_weights` attempts to load the default `rf-detr-large.pth` (1937 tokens, 44×44) into the new 1601-token model.
4. PyTorch raises `RuntimeError: Error(s) in loading state_dict ... size mismatch for ... position_embeddings: [1, 1937, 384] vs [1, 1601, 384]`.
5. `_should_fallback_to_deprecated_config` inspects the error: it's a `RuntimeError` containing both `"error(s) in loading state_dict"` and `"size mismatch"` → returns `True`.
6. `is_deprecated = True`, and `super().__init__` is retried — this time `get_model_config` returns `RFDETRLargeDeprecatedConfig` (patch=14, num_windows=4, PE=37).
7. The deprecated Large path happens to succeed because its default pretrained weights *do* fit its own PE grid.
8. A warning is logged (*"Automatically switched to deprecated model configuration, due to using deprecated weights"*) — but the message is misleading in this scenario, because the user didn't pass any deprecated weights. They passed a custom resolution on a modern model, and the new size-mismatch path from #956 was misclassified as "deprecated weights".
9. Construction returns successfully. `self.is_deprecated = True` but the user has no reason to check that flag — they wrote `RFDETRLarge(resolution=640)` and expected modern Large.
10. Later, during training, the backbone's stride check fires: deprecated Large has `patch_size × num_windows = 56`, and `640 % 56 ≠ 0`, so the forward pass throws the AssertionError above.

The user sees a stride-related error referencing a number (`56`) that has no obvious connection to anything they did. The warning about deprecated configuration was logged earlier, but it's easy to miss and its phrasing doesn't match what the user actually did.

I confirmed the fallback path is the one being taken with a second experiment — forcing `patch_size=16` explicitly:

```python
model = RFDETRLarge(resolution=640, patch_size=16)
```

```
RuntimeError: size mismatch for ... position_embeddings: [1, 1937, 384] vs [1, 1601, 384]
```

With `patch_size=16` pinned, the deprecated fallback would produce a `ValueError` with `"patch_size"` in the message during retry (deprecated Large requires 14), which the inner `except Exception: raise self.init_error` then unwraps — so we end up seeing the real underlying error from Problem A. That confirms the PE mismatch *is* what trips the fallback in the unforced case.

### Why the heuristic is now wrong

`_should_fallback_to_deprecated_config` was written under an assumption that held before #956: any `state_dict` size mismatch means the user is trying to load deprecated-format weights into a modern model. Post-#956, that assumption no longer holds. There is now a legitimate reason for a size mismatch that has nothing to do with deprecated weights — namely, a custom `resolution` that changes the modern model's PE grid relative to the modern pretrained checkpoint.

The heuristic can't distinguish between these two cases from the error message alone, because both produce the exact same `"size mismatch for ... position_embeddings"` `RuntimeError`. The only reliable way to tell them apart is to inspect *why* the target model has a different PE grid: if it's because the user passed a custom `resolution`, the weights aren't deprecated — they're the correct modern weights, and they just need to be interpolated to the new grid size.

### Proposed fix

The cleanest fix is to solve Problem A — once `load_pretrain_weights` does PE interpolation, custom resolutions on modern Large no longer cause a size mismatch, the heuristic stops firing incorrectly, and the deprecated fallback only triggers in its intended scenario (actually-deprecated weights).

A more defensive fix would additionally tighten `_should_fallback_to_deprecated_config` so that a size mismatch triggers fallback only when the user did *not* pass a custom `resolution` (or, more robustly, only when the checkpoint itself is identified as deprecated-format via metadata). That way, even if PE interpolation is added later, the fallback can't silently hijack unrelated errors.

Either way, when the fallback *does* fire, it would be much easier to debug if the log message included the captured `init_error` so users can see *why* the fallback was chosen, not just *that* it happened.


### Additional

_No response_

### Are you willing to submit a PR?

- [ ] Yes, I'd like to help by submitting a PR!

--- a/src/rfdetr/models/weights.py
+++ b/src/rfdetr/models/weights.py
@@ -42,9 +42,9 @@ _PE_KEY_SUFFIX = "embeddings.position_embeddings"
 
 def _interpolate_position_embeddings(
     checkpoint_state: dict,
-    model_config: ModelConfig,
+    pe_size: int,
 ) -> None:
-    """Interpolate DINOv2 positional embeddings in *checkpoint_state* to match *model_config*.
+    """Interpolate DINOv2 positional embeddings in *checkpoint_state* to match *pe_size*.
 
     When the model is configured with a custom ``resolution`` that differs from the
     checkpoint's training resolution, the DINOv2 backbone's ``position_embeddings``
@@ -52,16 +52,15 @@ def _interpolate_position_embeddings(
     skip shape mismatches on matching keys — it raises ``RuntimeError``.
 
     This function bicubic-interpolates every PE tensor in the checkpoint whose shape
-    differs from the target grid derived from ``model_config.positional_encoding_size``,
-    modifying *checkpoint_state* in-place before ``load_state_dict`` is called.
+    differs from the target grid, modifying *checkpoint_state* in-place before
+    ``load_state_dict`` is called.
 
     Args:
         checkpoint_state: The ``"model"`` sub-dict from a loaded checkpoint.
-        model_config: The model configuration used to derive the target PE size.
-            ``model_config.positional_encoding_size`` gives the target grid side length
-            (number of patches per spatial dimension, assuming a square grid).
+        pe_size: Target grid side length in patches (number of patches per spatial
+            dimension, assuming a square grid).  Typically
+            ``model_config.positional_encoding_size``.
     """
-    pe_size = model_config.positional_encoding_size
     n_target = pe_size * pe_size  # target number of patch tokens
 
     pe_keys = [k for k in checkpoint_state if k.endswith(_PE_KEY_SUFFIX)]
@@ -75,10 +74,8 @@ def _interpolate_position_embeddings(
         h_tgt = int(math.isqrt(n_target))
         if h_src * h_src != n_source or h_tgt * h_tgt != n_target:
             logger.warning(
-                "Skipping PE interpolation for %s: grid size is not a perfect square (source %d, target %d).",
-                key,
-                n_source,
-                n_target,
+                f"Skipping PE interpolation for {key}:"
+                f" grid size is not a perfect square (source {n_source}, target {n_target}).",
             )
             continue
 
@@ -262,7 +259,7 @@ def load_pretrain_weights(
         if any(name.endswith(x) for x in query_param_names):
             checkpoint["model"][name] = checkpoint["model"][name][:num_desired_queries]
 
-    _interpolate_position_embeddings(checkpoint["model"], mc)
+    _interpolate_position_embeddings(checkpoint["model"], mc.positional_encoding_size)
     nn_model.load_state_dict(checkpoint["model"], strict=False)
 
     # If the user explicitly set a class count larger than the checkpoint,

--- a/src/rfdetr/models/weights.py
+++ b/src/rfdetr/models/weights.py
@@ -19,11 +19,13 @@ extraction from ``detr.py:_load_pretrain_weights_into``.
 from __future__ import annotations
 
 import functools
+import math
 import os
 import warnings
 from typing import List
 
 import torch
+import torch.nn.functional as F  # noqa: N812
 
 from rfdetr.assets.model_weights import download_pretrain_weights, validate_pretrain_weights
 from rfdetr.config import ModelConfig, TrainConfig
@@ -34,6 +36,73 @@ from rfdetr.utilities.state_dict import _ckpt_args_get, validate_checkpoint_comp
 logger = get_logger()
 
 __all__ = ["load_pretrain_weights", "apply_lora"]
+
+_PE_KEY_SUFFIX = "embeddings.position_embeddings"
+
+
+def _interpolate_position_embeddings(
+    checkpoint_state: dict,
+    model_config: ModelConfig,
+) -> None:
+    """Interpolate DINOv2 positional embeddings in *checkpoint_state* to match *model_config*.
+
+    When the model is configured with a custom ``resolution`` that differs from the
+    checkpoint's training resolution, the DINOv2 backbone's ``position_embeddings``
+    parameter has an incompatible shape.  ``load_state_dict(strict=False)`` does **not**
+    skip shape mismatches on matching keys — it raises ``RuntimeError``.
+
+    This function bicubic-interpolates every PE tensor in the checkpoint whose shape
+    differs from the target grid derived from ``model_config.positional_encoding_size``,
+    modifying *checkpoint_state* in-place before ``load_state_dict`` is called.
+
+    Args:
+        checkpoint_state: The ``"model"`` sub-dict from a loaded checkpoint.
+        model_config: The model configuration used to derive the target PE size.
+            ``model_config.positional_encoding_size`` gives the target grid side length
+            (number of patches per spatial dimension, assuming a square grid).
+    """
+    pe_size = model_config.positional_encoding_size
+    n_target = pe_size * pe_size  # target number of patch tokens
+
+    pe_keys = [k for k in checkpoint_state if k.endswith(_PE_KEY_SUFFIX)]
+    for key in pe_keys:
+        ckpt_pe = checkpoint_state[key]  # [1, N_src+1, dim]
+        n_source = ckpt_pe.shape[1] - 1  # exclude class token
+        if n_source == n_target:
+            continue  # no mismatch — skip
+
+        h_src = int(math.isqrt(n_source))
+        h_tgt = int(math.isqrt(n_target))
+        if h_src * h_src != n_source or h_tgt * h_tgt != n_target:
+            logger.warning(
+                "Skipping PE interpolation for %s: grid size is not a perfect square (source %d, target %d).",
+                key,
+                n_source,
+                n_target,
+            )
+            continue
+
+        dim = ckpt_pe.shape[-1]
+        class_token = ckpt_pe[:, 0]  # [1, dim]
+        patch_pe = ckpt_pe[:, 1:]  # [1, N_src, dim]
+
+        patch_pe = patch_pe.reshape(1, h_src, h_src, dim).permute(0, 3, 1, 2)  # [1, dim, H, W]
+        patch_pe = F.interpolate(
+            patch_pe.float(),
+            size=(h_tgt, h_tgt),
+            mode="bicubic",
+            align_corners=False,
+            antialias=patch_pe.device.type != "mps",
+        ).to(ckpt_pe.dtype)
+        patch_pe = patch_pe.permute(0, 2, 3, 1).reshape(1, n_target, dim)  # [1, N_tgt, dim]
+
+        checkpoint_state[key] = torch.cat([class_token.unsqueeze(0), patch_pe], dim=1)
+        logger.debug(
+            "Interpolated positional embeddings %s: %s → %s.",
+            key,
+            tuple(ckpt_pe.shape),
+            tuple(checkpoint_state[key].shape),
+        )
 
 
 @deprecated(
@@ -193,6 +262,7 @@ def load_pretrain_weights(
         if any(name.endswith(x) for x in query_param_names):
             checkpoint["model"][name] = checkpoint["model"][name][:num_desired_queries]
 
+    _interpolate_position_embeddings(checkpoint["model"], mc)
     nn_model.load_state_dict(checkpoint["model"], strict=False)
 
     # If the user explicitly set a class count larger than the checkpoint,

--- a/src/rfdetr/models/weights.py
+++ b/src/rfdetr/models/weights.py
@@ -83,7 +83,7 @@ def _interpolate_position_embeddings(
             continue
 
         dim = ckpt_pe.shape[-1]
-        class_token = ckpt_pe[:, 0]  # [1, dim]
+        class_token = ckpt_pe[:, :1]  # [1, 1, dim] — keeps the sequence dimension
         patch_pe = ckpt_pe[:, 1:]  # [1, N_src, dim]
 
         patch_pe = patch_pe.reshape(1, h_src, h_src, dim).permute(0, 3, 1, 2)  # [1, dim, H, W]
@@ -96,7 +96,7 @@ def _interpolate_position_embeddings(
         ).to(ckpt_pe.dtype)
         patch_pe = patch_pe.permute(0, 2, 3, 1).reshape(1, n_target, dim)  # [1, N_tgt, dim]
 
-        checkpoint_state[key] = torch.cat([class_token.unsqueeze(0), patch_pe], dim=1)
+        checkpoint_state[key] = torch.cat([class_token, patch_pe], dim=1)
         logger.debug(
             "Interpolated positional embeddings %s: %s → %s.",
             key,

--- a/src/rfdetr/variants.py
+++ b/src/rfdetr/variants.py
@@ -144,7 +144,11 @@ class RFDETRLarge(RFDETR):
         # When the user explicitly sets a custom resolution, a PE size mismatch
         # is caused by the resolution change — not by deprecated weights.  Guard
         # against the fallback heuristic misclassifying it as deprecated weights.
-        _custom_resolution = "resolution" in kwargs
+        # Only suppress the fallback when the provided resolution genuinely differs
+        # from the class default; passing resolution=<default> explicitly (e.g. from
+        # a serialised config round-trip) must still allow the deprecated-weights retry.
+        _default_resolution = RFDETRLargeConfig.model_fields["resolution"].default
+        _custom_resolution = "resolution" in kwargs and kwargs.get("resolution") != _default_resolution
         try:
             super().__init__(**kwargs)
         except (ValueError, RuntimeError) as exc:

--- a/src/rfdetr/variants.py
+++ b/src/rfdetr/variants.py
@@ -141,10 +141,14 @@ class RFDETRLarge(RFDETR):
     def __init__(self, **kwargs):
         self.init_error = None
         self.is_deprecated = False
+        # When the user explicitly sets a custom resolution, a PE size mismatch
+        # is caused by the resolution change — not by deprecated weights.  Guard
+        # against the fallback heuristic misclassifying it as deprecated weights.
+        _custom_resolution = "resolution" in kwargs
         try:
             super().__init__(**kwargs)
         except (ValueError, RuntimeError) as exc:
-            if not self._should_fallback_to_deprecated_config(exc):
+            if _custom_resolution or not self._should_fallback_to_deprecated_config(exc):
                 raise
             self.init_error = exc
             self.is_deprecated = True

--- a/tests/training/test_detr_shim.py
+++ b/tests/training/test_detr_shim.py
@@ -1259,6 +1259,36 @@ class TestRFDETRLargeFallback:
         assert call_count == 2
         warn_spy.assert_called_once()
 
+    def test_pe_size_mismatch_with_custom_resolution_does_not_retry(self, monkeypatch, patch_lit):
+        """Custom resolution= must not trigger deprecated-config fallback on PE size mismatch.
+
+        Regression for #960: when ``resolution=`` is explicitly passed, a positional
+        embedding size mismatch is caused by the resolution change — not by deprecated
+        weights.  The fallback must be suppressed so the error surfaces to the caller
+        rather than silently loading the wrong model architecture.
+        """
+        call_count = 0
+
+        def _raise_pe_mismatch(self, **kwargs):
+            del self
+            nonlocal call_count
+            call_count += 1
+            raise RuntimeError(
+                "Error(s) in loading state_dict for LWDETR:\n\t"
+                "size mismatch for backbone.0.encoder.encoder.embeddings.position_embeddings: "
+                "copying a param with shape torch.Size([1, 577, 384]) from checkpoint, "
+                "the shape in current model is torch.Size([1, 1601, 384])."
+            )
+
+        monkeypatch.setattr(RFDETR, "__init__", _raise_pe_mismatch)
+
+        with pytest.raises(RuntimeError, match="size mismatch"):
+            RFDETRLarge(resolution=640)
+
+        assert call_count == 1, (
+            f"Expected no deprecated-config retry when resolution= is set, but __init__ was called {call_count} times."
+        )
+
 
 # ---------------------------------------------------------------------------
 # 7. _load_pretrain_weights_into — detr.py path (the non-PTL scenario from #806)

--- a/tests/training/test_load_pretrain_weights.py
+++ b/tests/training/test_load_pretrain_weights.py
@@ -368,6 +368,36 @@ class TestLoadPretrainWeightsPEInterpolation:
             "BaseConfig's non-formula-derived PE must be the interpolation target."
         )
 
+    def test_non_square_source_pe_logs_warning_and_is_not_modified(self, monkeypatch):
+        """Non-square source PE grids are skipped with a warning and left unchanged.
+
+        When ``n_source`` is not a perfect square the interpolation is skipped to
+        avoid producing malformed embeddings.  The tensor must remain untouched and
+        a warning must be emitted via the weights module logger.
+        """
+        mc = RFDETRNanoConfig(pretrain_weights="/fake/weights.pth", device="cpu")
+        # positional_encoding_size=24 → n_target=576 (perfect square, so the
+        # target-side guard does not trigger; only the source-side guard fires)
+
+        dim = 384
+        # 17 is not a perfect square: isqrt(17)=4, 4*4=16 ≠ 17
+        non_square_n_source = 17
+        original_pe = torch.randn(1, non_square_n_source + 1, dim)
+        checkpoint = _make_checkpoint(num_classes=91)
+        checkpoint["model"][PE_KEY] = original_pe.clone()
+
+        warning_calls: list[tuple] = []
+        monkeypatch.setattr("rfdetr.models.weights.logger.warning", lambda *a, **kw: warning_calls.append(a))
+        monkeypatch.setattr("rfdetr.models.weights.torch.load", lambda *a, **kw: checkpoint)
+        fake_model = MagicMock()
+        load_pretrain_weights(fake_model, mc)
+
+        pe = checkpoint["model"][PE_KEY]
+        assert torch.equal(pe, original_pe), "Non-square source PE must not be modified."
+        assert any("not a perfect square" in str(args) for args in warning_calls), (
+            f"Expected a 'not a perfect square' warning; got calls: {warning_calls}"
+        )
+
 
 # ---------------------------------------------------------------------------
 # Deprecation: train_config argument

--- a/tests/training/test_load_pretrain_weights.py
+++ b/tests/training/test_load_pretrain_weights.py
@@ -339,6 +339,34 @@ class TestLoadPretrainWeightsPEInterpolation:
         assert pe.shape == torch.Size([1, 577, dim]), "Matching PE shape must not be modified."
         assert torch.equal(pe, original_pe), "Matching PE tensor values must not be modified."
 
+    def test_base_config_non_formula_pe_is_interpolated_from_smaller_checkpoint(self, monkeypatch):
+        """RFDETRBaseConfig PE=37 (not formula-derived) is interpolated when checkpoint differs.
+
+        RFDETRBaseConfig.positional_encoding_size=37 is not updated by
+        ``_sync_pe_with_resolution`` because 37 ≠ 560//16=35 (not formula-derived).
+        Loading a checkpoint with a smaller PE grid (e.g., 24×24) must still
+        trigger interpolation to the model's fixed PE=37×37 target.
+        """
+        mc = RFDETRBaseConfig(pretrain_weights="/fake/weights.pth", device="cpu")
+        assert mc.positional_encoding_size == 37, "RFDETRBaseConfig PE must remain 37 (not formula-derived)"
+
+        dim = 384
+        src_pe_size = 24
+        src_n = src_pe_size * src_pe_size + 1
+        checkpoint = _make_checkpoint(num_classes=91)
+        checkpoint["model"][PE_KEY] = torch.randn(1, src_n, dim)
+
+        monkeypatch.setattr("rfdetr.models.weights.torch.load", lambda *a, **kw: checkpoint)
+        fake_model = MagicMock()
+        load_pretrain_weights(fake_model, mc)
+
+        pe = checkpoint["model"][PE_KEY]
+        expected_n = 37 * 37 + 1
+        assert pe.shape == torch.Size([1, expected_n, dim]), (
+            f"Expected PE shape [1, {expected_n}, {dim}] (37×37 grid), got {tuple(pe.shape)}. "
+            "BaseConfig's non-formula-derived PE must be the interpolation target."
+        )
+
 
 # ---------------------------------------------------------------------------
 # Deprecation: train_config argument

--- a/tests/training/test_load_pretrain_weights.py
+++ b/tests/training/test_load_pretrain_weights.py
@@ -337,6 +337,7 @@ class TestLoadPretrainWeightsPEInterpolation:
 
         pe = checkpoint["model"][PE_KEY]
         assert pe.shape == torch.Size([1, 577, dim]), "Matching PE shape must not be modified."
+        assert torch.equal(pe, original_pe), "Matching PE tensor values must not be modified."
 
 
 # ---------------------------------------------------------------------------

--- a/tests/training/test_load_pretrain_weights.py
+++ b/tests/training/test_load_pretrain_weights.py
@@ -252,6 +252,94 @@ class TestLoadPretrainWeightsSecondReinit:
 
 
 # ---------------------------------------------------------------------------
+# Regression #960: PE interpolation for custom resolution
+# ---------------------------------------------------------------------------
+
+PE_KEY = "backbone.0.encoder.encoder.embeddings.position_embeddings"
+
+
+class TestLoadPretrainWeightsPEInterpolation:
+    """Regression tests for #960 — PE must be interpolated when resolution changes.
+
+    ``load_pretrain_weights`` must bicubic-interpolate the checkpoint's DINOv2
+    positional embeddings to match the model's ``positional_encoding_size`` before
+    calling ``load_state_dict``.  Without this, any custom ``resolution`` that
+    changes the PE grid size causes a ``RuntimeError: size mismatch``.
+    """
+
+    @pytest.fixture(autouse=True)
+    def _patch_download(self, monkeypatch):
+        """Suppress all download and file-existence side effects."""
+        monkeypatch.setattr("rfdetr.models.weights.download_pretrain_weights", lambda *a, **kw: None)
+        monkeypatch.setattr("rfdetr.models.weights.validate_pretrain_weights", lambda *a, **kw: None)
+        monkeypatch.setattr("rfdetr.models.weights.validate_checkpoint_compatibility", lambda *a, **kw: None)
+        monkeypatch.setattr("rfdetr.models.weights.os.path.isfile", lambda _: True)
+
+    @pytest.mark.parametrize(
+        "src_pe_size, tgt_resolution, patch_size, expected_tgt_pe_size",
+        [
+            pytest.param(24, 640, 16, 40, id="nano_24x24_upscale_to_40x40"),
+            pytest.param(40, 384, 16, 24, id="nano_40x40_downscale_to_24x24"),
+            pytest.param(32, 640, 16, 40, id="small_32x32_upscale_to_40x40"),
+        ],
+    )
+    def test_pe_in_checkpoint_is_interpolated_to_model_resolution(
+        self, monkeypatch, src_pe_size, tgt_resolution, patch_size, expected_tgt_pe_size
+    ):
+        """Checkpoint PE is bicubic-interpolated to match model_config.positional_encoding_size.
+
+        Regression for #960: ``load_pretrain_weights`` must not raise ``RuntimeError``
+        when model resolution differs from checkpoint resolution.  The PE tensor in the
+        checkpoint must be resized in-place before ``load_state_dict`` is called.
+        """
+        mc = RFDETRNanoConfig(
+            pretrain_weights="/fake/weights.pth",
+            device="cpu",
+            resolution=tgt_resolution,
+            patch_size=patch_size,
+        )
+        assert mc.positional_encoding_size == tgt_resolution // patch_size
+
+        dim = 384
+        src_n = src_pe_size * src_pe_size + 1  # patches + class token
+        checkpoint = _make_checkpoint(num_classes=91)
+        checkpoint["model"][PE_KEY] = torch.randn(1, src_n, dim)
+
+        monkeypatch.setattr("rfdetr.models.weights.torch.load", lambda *a, **kw: checkpoint)
+        fake_model = MagicMock()
+        load_pretrain_weights(fake_model, mc)
+
+        pe = checkpoint["model"][PE_KEY]
+        expected_n = expected_tgt_pe_size * expected_tgt_pe_size + 1
+        assert pe.shape == torch.Size([1, expected_n, dim]), (
+            f"Expected PE shape [1, {expected_n}, {dim}], got {tuple(pe.shape)}. "
+            f"PE was not interpolated from {src_pe_size}x{src_pe_size} "
+            f"to {expected_tgt_pe_size}x{expected_tgt_pe_size}."
+        )
+
+    def test_matching_pe_shape_is_not_modified(self, monkeypatch):
+        """When checkpoint PE matches model expectations, the tensor is not changed.
+
+        Ensures PE interpolation is a no-op for same-resolution checkpoints so that
+        normal weight loading is unaffected.
+        """
+        mc = RFDETRNanoConfig(pretrain_weights="/fake/weights.pth", device="cpu")
+        # Default: positional_encoding_size=24 → PE = [1, 24*24+1, 384] = [1, 577, 384]
+
+        dim = 384
+        original_pe = torch.randn(1, 577, dim)
+        checkpoint = _make_checkpoint(num_classes=91)
+        checkpoint["model"][PE_KEY] = original_pe.clone()
+
+        monkeypatch.setattr("rfdetr.models.weights.torch.load", lambda *a, **kw: checkpoint)
+        fake_model = MagicMock()
+        load_pretrain_weights(fake_model, mc)
+
+        pe = checkpoint["model"][PE_KEY]
+        assert pe.shape == torch.Size([1, 577, dim]), "Matching PE shape must not be modified."
+
+
+# ---------------------------------------------------------------------------
 # Deprecation: train_config argument
 # ---------------------------------------------------------------------------
 

--- a/tests/training/test_load_pretrain_weights.py
+++ b/tests/training/test_load_pretrain_weights.py
@@ -303,7 +303,7 @@ class TestLoadPretrainWeightsPEInterpolation:
         dim = 384
         src_n = src_pe_size * src_pe_size + 1  # patches + class token
         checkpoint = _make_checkpoint(num_classes=91)
-        checkpoint["model"][PE_KEY] = torch.randn(1, src_n, dim)
+        checkpoint["model"][PE_KEY] = torch.randn(1, src_n, dim).half()  # float16 to verify dtype round-trip
 
         monkeypatch.setattr("rfdetr.models.weights.torch.load", lambda *a, **kw: checkpoint)
         fake_model = MagicMock()
@@ -316,6 +316,7 @@ class TestLoadPretrainWeightsPEInterpolation:
             f"PE was not interpolated from {src_pe_size}x{src_pe_size} "
             f"to {expected_tgt_pe_size}x{expected_tgt_pe_size}."
         )
+        assert pe.dtype == torch.float16, f"Dtype must be preserved after interpolation, got {pe.dtype}"
 
     def test_matching_pe_shape_is_not_modified(self, monkeypatch):
         """When checkpoint PE matches model expectations, the tensor is not changed.

